distribution/pull.go

```diff
@@ -183,3 +176,25 @@ func validateRepoName(name string) error {
 	}
 	return nil
 }
+
+type tmpFileWrapper struct {
```

Comment:
Based on usage of this wrapper I think it could fully wrap the `tmpFile` by exposing a Write method and providing a constructor which makes the tempfile. This could also protect calling a write after `didFirstRead` is true.